crypto/cipher_extra/e_chacha20poly1305.c

```diff
@@ -346,3 +382,317 @@ const EVP_AEAD *EVP_aead_chacha20_poly1305(void) {
 const EVP_AEAD *EVP_aead_xchacha20_poly1305(void) {
   return &aead_xchacha20_poly1305;
 }
+
+static int cipher_chacha20_poly1305_init_key(CIPHER_CHACHA_POLY_CTX *ctx,
+                                 const uint8_t user_key[CHACHA_KEY_LEN],
+                                 const uint8_t counter_nonce[CHACHA_CTR_IV_LEN])
+{
+  CIPHER_CHACHA_KEY *key = CC_KEY(ctx);
+  uint32_t i;
+  if (user_key) {
+    for (i = 0; i < (CHACHA_KEY_LEN / 4); i++) {
+      key->key[i] = *(((const uint32_t *) user_key) + i);
+    }
+  }
+  if (counter_nonce) {
+    for (i = 0; i < CHACHA_CTR_IV_LEN / 4; i++) {
+      key->counter_nonce[i] = *(((const uint32_t *) counter_nonce) + i);
+    }
+  }
+  key->partial_len = 0;
+  return 1;
+}
+
+static int cipher_chacha20_poly1305_init(EVP_CIPHER_CTX *ctx,
+                                         const uint8_t *key,
+                                         const uint8_t *iv, int32_t enc) {
+  CIPHER_CHACHA_POLY_CTX *cipher_ctx = CCP_CTX(ctx);
+  if (!key && !iv) {
+    return 1;
+  }
+  cipher_ctx->len.aad = 0;
+  cipher_ctx->len.text = 0;
+  cipher_ctx->pad_aad = 0;
+  cipher_ctx->poly_initialized = 0;
+  // Init can be called multiple times before starting the cipher to
+  // independently initialize any combination of Key/IV/NULL.
+  if (iv != NULL) {
+    // Start the counter at 0 and copy over the nonce(iv)
+    uint8_t counter_nonce[CHACHA_CTR_IV_LEN] = {0};
+    memcpy(counter_nonce + CHACHA_CTR_IV_LEN - CHACHA_IV_LEN, iv,
+           CHACHA_IV_LEN);
+    cipher_chacha20_poly1305_init_key(cipher_ctx, key, counter_nonce);
+    // Nonce occupies the last 3 indices of the array
+    cipher_ctx->iv[0] = cipher_ctx->key.counter_nonce[1];
+    cipher_ctx->iv[1] = cipher_ctx->key.counter_nonce[2];
+    cipher_ctx->iv[2] = cipher_ctx->key.counter_nonce[3];
+  } else {
+    cipher_chacha20_poly1305_init_key(cipher_ctx, key, NULL);
+  }
+  return 1;
+}
+
+static int cipher_chacha20_do_cipher(EVP_CIPHER_CTX *ctx, uint8_t *out,
+                                     const uint8_t *inp, size_t len)
+{
+  CIPHER_CHACHA_POLY_CTX  *cipher_ctx = CCP_CTX(ctx);
+  CIPHER_CHACHA_KEY *key = CC_KEY(cipher_ctx);
+  uint32_t n, rem, ctr32;
+
+  // Complete any partial block
+  if ((n = key->partial_len)) {
+    // Compute the cipher using our partially used key and any new input up
+    // to the next block
+    while (len && n < CHACHA_BLOCK_LEN) {
+      *out++ = *inp++ ^ key->buf[n++];
+      len--;
+    }
+    key->partial_len = n;
+
+    // If we consumed all the input, we're done
+    if (len == 0) {
+      return 1;
+    }
+
+    // If we completed a block, increment the counter
+    if (n == CHACHA_BLOCK_LEN) {
+      key->partial_len = 0;
+      key->counter_nonce[0]++;
```

Comment:
Is this sufficient? Isn't the counter_nonce 4 u32 long? What happens if 0 overflows?

Reply:
ChaCha20-Poly1305 is defined in RFC 8439 and the limits on the algorithm are defined in [section 2.8]
(https://datatracker.ietf.org/doc/html/rfc8439#section-2.8). They merely state the limits but do not define what should happen if we exceed them. It's up to us to define that. OpenSSL and BoringSSL have chosen to wrap the counter so I'll follow that approach as there's no conventional way of erroring out here.

>1.  The amount of encrypted data possible in a single invocation is
> 2^32-1 blocks of 64 bytes each, because of the size of the block
> counter field in the ChaCha20 block function.  This gives a total
> of 274,877,906,880 bytes, or nearly 256 GB.  This should be
> enough for traffic protocols such as IPsec and TLS, but may be
> too small for file and/or disk encryption.  For such uses, we can
> return to the original design, reduce the nonce to 64 bits, and
> use the integer at position 13 as the top 32 bits of a 64-bit
> block counter, increasing the total message size to over a
> million petabytes (1,180,591,620,717,411,303,360 bytes to be
> exact).
>
>
> 2.  Despite the previous item, the ciphertext length field in the
> construction of the buffer on which Poly1305 runs limits the
> ciphertext (and hence, the plaintext) size to 2^64 bytes, or
> sixteen thousand petabytes (18,446,744,073,709,551,616 bytes to
> be exact).

Reply:
I read through some of the code to see what our AES CTR impl. does when exposed through the EVP_CIPHER interface. It looks like [the cipher update function](https://github.com/aws/aws-lc/blob/f3e49b01fb9a58bddd5bac218d9369e36d5bfbd0/crypto/fipsmodule/cipher/e_aes.c#L276) has _no error path_ and thus erroring on counter overflow would be impossible for that function (however, other AES impls. [do appear to guard against overflow](https://github.com/aws/aws-lc/blob/f3e49b01fb9a58bddd5bac218d9369e36d5bfbd0/crypto/fipsmodule/cipher/e_aes.c#L1499)).

anyway, since other relevant EVP_CIPHERs don't perform this check and it's not mandated by the RFC, I don't think this is a blocking concern. it might be a useful security feature to be added to all counter-based EVP_CIPHERs in the future.